server/civitas/services/UserService.js

```diff
@@ -0,0 +1,453 @@
+        {
+          statusCode: 500,
+          body: err
+        },
+        userChannels.Internal.DeleteCompletedEvent
+      );
+    }
+
+    return subscriptionManager.emitInternalResponseEvent(
+      {
+        statusCode: 200,
+        body: user
+      },
+      userChannels.Internal.DeleteCompletedEvent
+    );
+  },
+
+  /**
+   * Updates a user
+   *
+   * @param {object} request - The request arguments passed in from the controller
+   */
+  updateUser: async function updateUser(request) {
+
+    let user = null;
+
+    try {
+      user = await User.findById(request.id);
+    } catch (err) {
+      return subscriptionManager.emitInternalResponseEvent(
+        {
+          statusCode: 500,
+          body: err
+        },
+        userChannels.Internal.UpdateCompletedEvent
+      );
+    }
+
+    var validationResult = await userValidator.validateUpdate(user, request);
+
+    if (validationResult) {
+      return subscriptionManager.emitInternalResponseEvent(
+        {
+          statusCode: 400,
+          body: validationResult
+        },
+        userChannels.Internal.CreateCompletedEvent
+      );
+    }
+
+    logging.logAction(
+      logging.logLevels.INFO,
+      `Attempting to update a user document with id [${user.id}]`
+    );
+
+    let userAddresses = null;
+
+    // First update the address
+    if (request.addresses) {
+
+      try {
+        userAddresses = await _updateUserAddress(user, request);
+      } catch (error) {
+
+        let statusCode = 0;
+
+        if (error.status === 400) {
+          statusCode = 400;
+        } else {
+          statusCode = 500;
+        }
+
+        return subscriptionManager.emitInternalResponseEvent(
+          {
+            statusCode: statusCode,
+            body: error
+          },
+          userChannels.Internal.UpdateCompletedEvent
+        );
+      }
+
+      user.addresses = userAddresses.map((userAddress) => {
+        return userAddress.id;
+      });
+    }
+
+    if (request.firstName) {
+      user.firstName = request.firstName;
+    }
+
+    if (request.lastName) {
+      user.firstName = request.lastName;
+    }
+
+    if (request.email) {
+      user.email = request.email;
+    }
+    
+    try {
+      await user.save();
+    } catch (err) {
+      return subscriptionManager.emitInternalResponseEvent(
+        {
+          statusCode: 500,
+          body: err
+        },
+        userChannels.Internal.UpdateCompletedEvent
+      );
+    }
+
+    user.updatedAt = new Date();
+
+    // set the address objects back on the display response
+    user.addresses = userAddresses;
+    return subscriptionManager.emitInternalResponseEvent(
+      {
+        statusCode: 200,
+        body: user
+      },
+      userChannels.Internal.UpdateCompletedEvent
+    );
+  }
+};
+
+function _updateUserAddress(user, request) {
+
+  return new Promise(async function (resolve, reject) {
+    // It is easier just to remove the existing addresses and insert the new ones
```

Comment:
it might create problems, since the Id will changes, if we wanna use that specific address somewhere in our application, we need a way to replace the used address

Reply:
I thought about it, it is easier this way, if it becomes a problem, which I don't think, then we can change the way it works